### ISSUE TITLE
hard negative mining based on similarity score

### ISSUE THREAD

**Issue by daegonYu** [FlagOpen/FlagEmbedding#1130]
In FlagEmbedding, hard negative mining is extracted based on ranking. (FlagEmbedding/baai_general_embedding/finetune/hn_mine.py) Is there a code that does hard negative mining based on similarity score?

**Comment by staoxiao** [comment]
Sorry, the script doesn't support this function. But you can add this function easily in hn_mine.py

**Comment by daegonYu** [comment]
While waiting for an answer, I wrote a code and it seems that extracting hard negatives by cosine similarity score is more useful than extracting hard negatives by rank. This is also presented in the paper https://arxiv.org/pdf/2405.05374 (Appendix Algorithm 1: Tunable Negative Mining). Can I send a PR so that others can use it? The existence of the topk argument has the advantage of allowing both rank and cosine similarity scores to be considered.

FlagEmbedding/FlagEmbedding/baai_general_embedding/finetune/hn_mine_by_similarity.py

```py

import argparse
import json
import random
import numpy as np
import faiss
from tqdm import tqdm

from FlagEmbedding import FlagModel
from FlagEmbedding.baai_general_embedding.finetune.hn_mine import create_index, batch_search, get_corpus

def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('--model_name_or_path', default="BAAI/BGE-m3", type=str)
    parser.add_argument('--input_file', default=None, type=str)
    parser.add_argument('--candidate_pool', default=None, type=str)
    parser.add_argument('--output_file', default=None, type=str)
    parser.add_argument('--similarity_range', default="0.5-0.8", type=str, help="similarity range to sample negatives")
    parser.add_argument('--use_gpu_for_searching', action='store_true', help='use faiss-gpu')
    parser.add_argument('--negative_number', default=15, type=int, help='the number of negatives')
    parser.add_argument('--topk', default=200, type=int, help='number of top k candidates to consider')
    parser.add_argument('--query_instruction_for_retrieval', default="")

    return parser.parse_args()


def find_knn_neg(model, input_file, candidate_pool, output_file, similarity_range, negative_number, use_gpu, topk):
    corpus = []
    queries = []
    train_data = []
    for line in open(input_file):
        line = json.loads(line.strip())
        train_data.append(line)
        corpus.extend(line['pos'])
        if 'neg' in line:
            corpus.extend(line['neg'])
        queries.append(line['query'])

    if candidate_pool is not None:
        if not isinstance(candidate_pool, list):
            candidate_pool = get_corpus(candidate_pool)
        corpus = list(set(candidate_pool))
    else:
        corpus = list(set(corpus))

    print(f'inferencing embedding for corpus (number={len(corpus)})--------------')
    p_vecs = model.encode(corpus, batch_size=256)
    print(f'inferencing embedding for queries (number={len(queries)})--------------')
    q_vecs = model.encode_queries(queries, batch_size=256)

    print('create index and search------------------')
    index = create_index(p_vecs, use_gpu=use_gpu)
    all_scores, all_inxs = batch_search(index, q_vecs, topk=topk) 

    min_sim, max_sim = map(float, similarity_range)

    for i, data in enumerate(train_data):
        query = data['query']
        scores = all_scores[i]
        inxs = all_inxs[i]
        
        filtered_inx = []
        for score, inx in zip(scores, inxs):
            if min_sim <= score <= max_sim and corpus[inx] not in data['pos'] and corpus[inx] != query:
                filtered_inx.append(inx)

        if len(filtered_inx) > negative_number:
            filtered_inx = random.sample(filtered_inx, negative_number)
        data['neg'] = [corpus[inx] for inx in filtered_inx]

    with open(output_file, 'w') as f:
        for data in train_data:
            if len(data['neg']) < negative_number:
                samples = random.sample(corpus, negative_number - len(data['neg']) + len(data['pos']))
                samples = [sent for sent in samples if sent not in data['pos']]
                data['neg'].extend(samples[: negative_number - len(data['neg'])])
            f.write(json.dumps(data, ensure_ascii=False) + '\n')


if __name__ == '__main__':
    args = get_args()
    similarity_range = args.similarity_range.split('-')
    similarity_range = [float(x) for x in similarity_range]

    model = FlagModel(args.model_name_or_path, query_instruction_for_retrieval=args.query_instruction_for_retrieval)

    find_knn_neg(model,
                 input_file=args.input_file,
                 candidate_pool=args.candidate_pool,
                 output_file=args.output_file,
                 similarity_range=similarity_range,
                 negative_number=args.negative_number,
                 use_gpu=args.use_gpu_for_searching,
                 topk=args.topk)

**Comment by staoxiao** [comment]
Sure, we warmly welcome contributions via PR.  I suggest setting the default value to 0-1,  allowing users to adjust this parameter themselves.

**Comment by daegonYu** [comment]
I have submitted a pull request. As you said, I have set the default value to 0-1.

#1143

**Comment by maosl2000** [comment]
请问如何验证挖掘的数据质量呢 我不太懂关于数据质量的指标

**Comment by daegonYu** [comment]
@maosl2000 

Well, I train it, look at the performance of the model, and evaluate the quality of the training data. I understand that appropriate data may vary depending on the task or domain.
